src/canopy/knowledge_base/record_encoder/hybrid.py

```diff
@@ -0,0 +1,122 @@
+
+        super().__init__(**kwargs)
+        if dense_encoder:
+            if not isinstance(dense_encoder, BaseDenseEncoder):
+                raise TypeError(
+                    f"dense_encoder must be an instance of BaseDenseEncoder, "
+                    f"not {type(dense_encoder)}"
+                )
+            self._dense_encoder = dense_encoder
+        else:
+            self._dense_encoder = OpenAIEncoder()
+        self._sparse_encoder = BM25Encoder.default()
+        self._alpha = alpha
+
+    def _encode_documents_batch(self,
+                                documents: List[KBDocChunk]
+                                ) -> List[KBEncodedDocChunk]:
+        """
+        Encode a batch of documents, takes a list of KBDocChunk and returns a list of KBEncodedDocChunk.
+
+        Args:
+            documents: A list of KBDocChunk to encode.
+        Returns:
+            encoded chunks: A list of KBEncodedDocChunk,
+            with the `values` containing the generated dense vector and
+            `sparse_values` containing the generated sparse vector.
+        """  # noqa: E501
+        texts = [d.text for d in documents]
+        dense_values = self._dense_encoder.encode_documents(texts)
+        sparse_values = self._sparse_encoder.encode_documents(texts)
+        return [
+            KBEncodedDocChunk(**d.dict(), values=v, sparse_values=sv) for d, v, sv in
+            zip(documents, dense_values, sparse_values)
+        ]
+
+    def _encode_queries_batch(self, queries: List[Query]) -> List[KBQuery]:
+        """
+        Encode a batch of queries, takes a list of Query and returns a list of KBQuery.
+        Args:
+            queries: A list of Query to encode.
+        Returns:
+            encoded queries: A list of KBQuery, with the `values` containing the generated dense vector with the weight
+            alpha and `sparse_values` containing the generated sparse vector with the weight (1 - alpha).
+        """  # noqa: E501
+
+        texts = [q.text for q in queries]
+        dense_values = self._dense_encoder.encode_queries(texts)
+        sparse_values = self._sparse_encoder.encode_queries(texts)
+        scaled_values = [
+            hybrid_convex_scale(v, sv, self._alpha) for v, sv in
+            zip(dense_values, sparse_values)
+        ]
+
+        return [KBQuery(**q.dict(), values=v, sparse_values=sv) for q, (v, sv) in
+                zip(queries, scaled_values)]
+
+    @cached_property
+    def dimension(self) -> int:
+        """
+        The dimension is the length of the vector generated by the `DenseRecordEncoder`
+        Canopy will run a single word through the encoder to get the dimension, this will also validate that the encoder
+        is working properly.
+
+        Returns:
+            dimension(int): the dimension of the encoder
+        """  # noqa: E501
+        dummy_doc = KBDocChunk(text="hello", id="dummy_doc", document_id="dummy_doc")
```

Comment:
You can simply call `return self.dense_encoder.dimension`.  
That's better, since in some of the dense encoders we wrap that call with proper error handling etc. Not a good idea to re-implement here.